internal/declcfg/diff.go

```diff
@@ -0,0 +1,393 @@
+package declcfg
+
+import (
+	"reflect"
+	"sort"
+
+	"github.com/blang/semver"
+	"github.com/mitchellh/hashstructure/v2"
+
+	"github.com/operator-framework/operator-registry/internal/model"
+	"github.com/operator-framework/operator-registry/internal/property"
+)
+
+// Diff returns a Model containing everything in newModel not in oldModel,
+// and all bundles that exist in oldModel but are different in newModel.
+// If oldModel is empty, only channel heads in newModel's packages are
+// added to the output Model. All dependencies not in oldModel are also added.
+func Diff(oldModel, newModel model.Model) (model.Model, error) {
```

Comment:
As things stand now, this makes sense. However the signature might change (follow-up) to accept an iterator type that can block-load packages from disk to overcome memory limitations. I'm not 100% sure where that iterator's definition will live yet, but it makes sense to place it in `declcfg` because it'll index packages on-disk at parse time.